doc/source/ray-observability/monitoring-debugging/troubleshoot-failures.rst

```diff
@@ -1,44 +1,252 @@
+If Ray workers access the object inside the object store using ``ray.get``, SHR usage increases. Since the Ray object store supports the :ref:`zero-copy <serialization-guide>`
+deserialization, several workers can access the same object without copying them to in-process memory. For example, if
+8 workers access the same object inside the Ray object store, each process' ``SHR`` usage increases. However, they are not using 8 * SHR memory (there's only 1 copy in the shared memory). 
+Also note that Ray object store triggers :ref:`object spilling <object-spilling>` when the object usage goes beyond the limit, which means the memory usage from the shared memory won't exceed 30%
+of the host memory.
+
+It means when there are out-of-memory issues from a host, it is due to RSS usage from each worker. We recommend to calculate per
+process memory usage by RSS - SHR because SHR is for Ray object store as explained above. That said, the total memory usage is typically
+``SHR (object store memory usage, 30% of memory) + sum(RSS - SHR from each ray proc) + sum(RSS - SHR from system components. e.g., raylet, GCS. Usually small)``.
+
+**Step 2 Debug memory problems, head node**
+
+First, check the head node memory usage from the metrics page. Find the head node address from the cluster page.
+
+.. image:: ../images/head-node-addr.png
+    :align: center
+
+And then check the memory usage from the head node from the node memory usage view inside the Dashboard :ref:`metrics view <dash-metrics-view>`.
+
+.. image:: ../images/metrics-node-view.png
+    :align: center
+
+Ray head node has more memory-demanding system components such as GCS or the dashboard. 
+Also, driver runs from a head node by default. If the head node has the same memory capacity as worker nodes
+and if you execute the same number of tasks and actors from a head node, it can easily have out-of-memory problems.
+In this case, we recommend you to not run any tasks and actors to the head node. This is achievable by specifying ``--num-cpus=0`` when starting a head node by ``ray start --head``.
+
+**Step 3 Debug memory problems, parallelism**
+
+High parallelism can trigger out-of-memory errors. For example, imagine 
+you have 8 training workers that perform the data preprocessing -> training. 
+If you load too much data into each worker, the total memory usage (``training worker mem usage * 8``) can exceed the
+memory capacity. 
+
+You can verify it by looking at the :ref:`per task and actor memory usage graph <dash-workflow-cpu-memory-analysis>`. And the task metrics.
+
+First, see the memory usage of a ``allocate_memory`` task. It is total 21GB.
+At the same time, you can verify 15 concurrent tasks running.
+
+.. image:: ../images/component-memory.png
+    :align: center
+
+.. image:: ../images/tasks-graph.png
+    :align: center
+
+It means each task uses about 21GB / 15 == 1.4 GB. You can increase the ``num_cpus`` requirement of each task to reduce the parallelism to solve the problem.
```

Comment:
You can link to https://docs.ray.io/en/latest/ray-core/patterns/limit-running-tasks.html